src/wasm.cc

```diff
@@ -606,9 +595,9 @@ std::shared_ptr<PluginHandleBase> getOrCreateThreadLocalPlugin(
     if (plugin_handle) {
       return plugin_handle;
     }
-    // Remove stale entry.
-    local_plugins.erase(key);
   }
+  // Remove stale entries.
+  cleanupLocalCache(local_plugins);
```

Comment:
This might be quadratic in the number of active VMs. Maybe try to rate limit it - run once it grows x2 or with a timer?

Reply:
I agree, but my intuition is that the number of plugins should be small enough to make cleanup time noticeable (and, for example, Envoy runs each VM in its own thread, AFAIK.) Also this should be a slow execution path anyway since we got a cache miss and need to initialize the plugin (and possibly VM) later.
Although we could trigger cleanup by cache size or by time-stamping last cleanup with a monotonic clock, any hardcoded threshold/period seems (too) arbitrary to me.

Reply:
> the number of plugins should be small enough to make cleanup time noticeable

While that's probably true for the number of unique Proxy-Wasm plugins, the number of VMs (e.g. when using `vm_id`) and plugin instances (e.g. when running thousands of vhosts with different plugin configuration each) can be significantly higher.

I'm not sure how big of an issue this would be in practice, but it could quickly get out of hand.

Maybe add a "possibly expired" list that's populated in destructors?

> Envoy runs each VM on its own thread, AFAIK.

For each Proxy-Wasm plugin, Envoy starts a VM on each of Envoy's worker threads. It doesn't create a new thread for that purpose, although, technically, some Wasm engines could do that underneath.

Reply:
> For each Proxy-Wasm plugin, Envoy starts a VM on each of Envoy's worker threads. It doesn't create a new thread for that purpose, although, technically, some Wasm engines could do that underneath.

Ah, I see, I mistook a "worker" for thread.

Reply:
> Maybe add a "possibly expired" list that's populated in destructors?

I thought about placing something like that in the std::shared_ptr's deleter, but VM/plugin factory is passed by argument to getOrCreateThreadLocalWasm/getOrCreateThreadLocalPlugin, so their creation is opaque for those functions.
We could inject callbacks into WasmHandleBase/PluginHandleBase destructors, which might be a bit kludgy without changing constructors (and library API.)

Reply:
If we have the Wasm*Base destructors clear entries from the thread-local caches, it could be empty work for instances that were not created via a getOrCreateThreadLocal* function.

Another option would be to add threadLocalWasmDeleted() and threadLocalPluginDeleted() functions and leave it up to the caller of getOrCreateThreadLocal* to call, perhaps from the destructors of subclasses of WasmHandleBase/PluginHandleBase. That's more work for users of getOrCreateThreadLocal*, though.

One other idea: every time getOrCreateThreadLocalWasm() is called, have it check for expiry of up to N entries in the map, where N is some small bounded number (say, 5). To avoid checking the same N entries each time, you could maintain a deque alongside the unordered_map, to keep track of which entries to reap next.

Reply:
> If we have the Wasm_Base destructors clear entries from the thread-local caches, it could be empty work for instances that were not created via a getOrCreateThreadLocal_ function.

I was thinking about some optional std::function callback member and a companion function like `onDestroy` to set this callback. This way we could call it in `getOrCreateThreadLocalWasm` or `getOrCreateThreadLocalPlugin` and wouldn't impose additional work in other cases. It's not what I would call a clean design but it should work.

> One other idea: every time getOrCreateThreadLocalWasm() is called, have it check for expiry of up to N entries in the map, where N is some small bounded number (say, 5). To avoid checking the same N entries each time, you could maintain a deque alongside the unordered_map, to keep track of which entries to reap next.

I like the idea of having a couple of deques (with `(key, weak ptr)` tuples) and checking no more than N entries at a time. If I understood it correctly, we expect problems mostly with plugin cache due to its cardinality which is roughly N_vms * M_vm_plugins. What if we make a 2-level cache? First we narrow the search by a VM key and then by a plugin key. VMs plugin set should have a manageable size. What do you think?

Reply:
> I was thinking about some optional std::function callback member and a companion function like `onDestroy` to set this callback. This way we could call it in `getOrCreateThreadLocalWasm` or `getOrCreateThreadLocalPlugin` and wouldn't impose additional work in other cases. It's not what I would call a clean design but it should work.

That sounds reasonable to me. I agree it's not super-elegant, but it's straightforward and doesn't waste cycles probing VMs that are still live. You would need to add setter methods to both WasmBase and PluginHandleBase, right? For naming consistency I'd suggest calling it something like `addOnDestroyAction` or `addCleanupActio`n (or "Callback" instead of "Action"), similar to [`WasmBase::addAfterVmCallAction`](https://github.com/proxy-wasm/proxy-wasm-cpp-host/blob/c932df12c9db59d331dcb1020174d5f911f67244/include/proxy-wasm/wasm.h#L146) or [`WasmVm::addFailCallback`](https://github.com/proxy-wasm/proxy-wasm-cpp-host/blob/c932df12c9db59d331dcb1020174d5f911f67244/include/proxy-wasm/wasm_vm.h#L320).

(Calling it `onDestroy` would resemble proxy-wasm callbacks like `onDone` or `onDelete`; even though those are on Context objects, the similar naming could invite confusion.)

> > One other idea: every time getOrCreateThreadLocalWasm() is called, have it check for expiry of up to N entries in the map, where N is some small bounded number (say, 5). To avoid checking the same N entries each time, you could maintain a deque alongside the unordered_map, to keep track of which entries to reap next.
> 
> I like the idea of having a couple of deques (with `(key, weak ptr)` tuples) and checking no more than N entries at a time. If I understood it correctly, we expect problems mostly with plugin cache due to its cardinality which is roughly N_vms * M_vm_plugins. What if we make a 2-level cache? First we narrow the search by a VM key and then by a plugin key. A single VM plugin set should have a manageable size. What do you think?

If you go this route, I'd propose keeping it as simple as possible and just having a single-level deque that you rotate through. But I think the destroy callback seems simpler, so would lean towards that even though it adds a couple methods.

Reply:
I opted for a solution with queues, as it seems to me the most obvious and encapsulated behind the current interface.

Reply:
Nice!